manifests/conf.pp

```diff
@@ -4,21 +4,15 @@
 
 class device_manager::conf {
 
-  if ($facts['os']['family'] != 'windows') {
-    File {
-      owner => $::identity['user'],
-      group => $::identity['group'],
-      mode  => '0640',
-    }
-  }
+  # Cannot pass parameters to included classes.
+  $owner = lookup('device_manager::conf::owner', {'default_value' => $::identity['user']})
+  $group = lookup('device_manager::conf::group', {'default_value' => $::identity['group']})
```

Comment:
Please follow the recommendation from the Style Guide here: https://puppet.com/docs/puppet/latest/style_guide.html#functions

Reply:
If you mean:

> You should avoid using calls to Hiera functions in modules meant for public consumption, because not all users have implemented Hiera. Instead, we recommend using parameters that can be overridden with Hiera.

... then I need to implement this as a parameter to the defined type rather than in Hiera as a parameter to the class, and not support configuring a user in versions of Puppet (pre 5.0.0) that do not support `puppet device --target`. 

In fact, it may be time to drop compatibility for pre 5.0.0 Puppet and drop the "untargeted" code, given other components (like cisco_ios) require Puppet 5.0.0+ ?

Reply:
Here is what that looks like:

https://github.com/puppetlabs/device_manager/compare/master...tkishel:MODULES-9191_support_configurable_user_v2?expand=1

Reply:
That is one option. You don't have to go that far though. Adding the `owner` and `group` params to the `device::conf` class would automatically do the hiera lookup you explicitly coded:

```
class device::conf($owner = $::identity['user'], ...) {
```

I'm not sure if the default is evaluated correctly. If that causes the hiera lookup to be skipped, the other pattern would be:

```
class device::conf($owner, ...) {
  $real_owner = $owner ? { undef => $::identity['user'], default => $owner }
```

This behaves exactly the same as your version for hiera users, but non-hiera-users can manually declare the `device::conf` class before everything else in their profile with the values specified manually.

Reply:
Alas, `device_manager::conf` is Include-like (potentially included by multiple `device_manager` resources) so it cannot be called with parameters. I've update the other option and will make a formal PR for it.

Reply:
You're one of today's lucky 10k. Folks not using hiera are usually using the pattern as exemplified below. This relies on the parse order of manifests, which can be brittle, but having the resources together in a profile class alleviates this.

For each class, there can be one resource-like declaration and an arbitrary number of `include`s, as long as the resource-like declaration is parsed first.

```
david@davids:~/git/device_manager$ git diff manifests/conf.pp
diff --git a/manifests/conf.pp b/manifests/conf.pp
index 1301a5e..265f6bc 100644
--- a/manifests/conf.pp
+++ b/manifests/conf.pp
@@ -2,12 +2,10 @@
 # Using concat instead of inifile, because purgable.
 # @api private
 
-class device_manager::conf {
-
-  # Cannot pass parameters to included classes.
-  $owner = lookup('device_manager::conf::owner', {'default_value' => $::identity['user']})
-  $group = lookup('device_manager::conf::group', {'default_value' => $::identity['group']})
-
+class device_manager::conf(
+  $owner = $::identity['user'],
+  $group = $::identity['group'])
+{
   # Use a fact to identify the confdir file on this agent.
   $devices_directory = "${::puppet_settings_confdir}/puppet/devices"
 
david@davids:~/git/device_manager$ cat test.pp 
class {
  'device_manager::conf':
    owner => david,
    group => david
}


device_manager {
  'test':
    type        => 'foo',
    credentials => {
      user => 'admin'
    }
}
david@davids:~/git/device_manager$ pdk bundle exec puppet apply --modulepath spec/fixtures/modules/ --noop test.pp 
pdk (INFO): Using Ruby 2.5.3
pdk (INFO): Using Puppet 6.10.0
Notice: Compiled catalog for localhost in environment production in 0.27 seconds
Notice: /Stage[main]/Device_manager::Conf/File[/home/david/.puppetlabs/etc/puppet/devices/the_node]/ensure: current_value 'directory', should be 'absent' (noop)
Notice: /home/david/.puppetlabs/etc/puppet/devices: Would have triggered 'refresh' from 1 event
Notice: /Stage[main]/Device_manager::Conf/Concat[/home/david/.puppetlabs/etc/puppet/device.conf]/File[/home/david/.puppetlabs/etc/puppet/device.conf]/content: current_value '{md5}58eccb355f436b1bc5aacd7f022e1a15', should be '{md5}6693f2098352324aa577205bfa77850b' (noop)
Notice: /Stage[main]/Device_manager::Conf/Concat[/home/david/.puppetlabs/etc/puppet/device.conf]/File[/home/david/.puppetlabs/etc/puppet/device.conf]/mode: current_value '0644', should be '0640' (noop)
Notice: /home/david/.puppetlabs/etc/puppet/device.conf: Would have triggered 'refresh' from 2 events
Notice: Concat[/home/david/.puppetlabs/etc/puppet/device.conf]: Would have triggered 'refresh' from 3 events
Notice: Class[Device_manager::Conf]: Would have triggered 'refresh' from 2 events
Notice: /Stage[main]/Main/Device_manager[test]/Device_manager::Conf::Device[test]/File[/home/david/.puppetlabs/etc/puppet/devices/test.conf]/ensure: current_value 'absent', should be 'file' (noop)
Notice: /Stage[main]/Device_manager::Fact/File[/home/david/.puppetlabs/etc/facter]/ensure: current_value 'absent', should be 'directory' (noop)
Notice: /Stage[main]/Device_manager::Fact/File[/home/david/.puppetlabs/etc/facter/facts.d]/ensure: current_value 'absent', should be 'directory' (noop)
Notice: Device_manager::Conf::Device[test]: Would have triggered 'refresh' from 1 event
Notice: /Stage[main]/Device_manager::Fact/Concat[/home/david/.puppetlabs/etc/facter/facts.d/devices.yaml]/File[/home/david/.puppetlabs/etc/facter/facts.d/devices.yaml]/ensure: current_value 'absent', should be 'file' (noop)
Notice: /home/david/.puppetlabs/etc/facter/facts.d/devices.yaml: Would have triggered 'refresh' from 1 event
Notice: Concat[/home/david/.puppetlabs/etc/facter/facts.d/devices.yaml]: Would have triggered 'refresh' from 2 events
Notice: Class[Device_manager::Fact]: Would have triggered 'refresh' from 3 events
Notice: Device_manager[test]: Would have triggered 'refresh' from 1 event
Notice: Class[Main]: Would have triggered 'refresh' from 1 event
Notice: Stage[main]: Would have triggered 'refresh' from 3 events
Notice: Applied catalog in 0.09 seconds
david@davids:~/git/device_manager$ 
```

Filed under https://xkcd.com/1053/

Reply:
That pattern is clever! But it, or having some of device configuration configured via resource parameters and some configured via class parameters seems awkward if not magical. I would rather degrade "gracefully" for pre Puppet 5.x as per the alternative PR ... but if you feel strongly about this, I can certainly modify this PR to incorporate this pattern.